token/cli/src/main.rs

```diff
@@ -702,24 +703,34 @@ fn command_transfer(
     let mut recipient_token_account = recipient;
     let mut minimum_balance_for_rent_exemption = 0;
 
-    let recipient_is_token_account = if !config.sign_only {
+    let recipient_is_token_account = if !config.sign_only{
         let recipient_account_info = config
             .rpc_client
-            .get_account_with_commitment(&recipient, config.rpc_client.commitment())?
-            .value
-            .map(|account| {
-                account.owner == config.program_id && account.data.len() == Account::LEN
-            });
-
-        if recipient_account_info.is_none() && !allow_unfunded_recipient {
+            .get_account_with_commitment(&recipient,config.rpc_client.commitment())?
+            .value;
+        if let Some(account) = recipient_account_info {
+            if account.owner == config.program_id && account.data.len() == Account::LEN {
+                true
+            } 
+            else if account.owner != system_program::id() && !allow_non_system_account_recipient {
+                return Err("Error: The recipient address is not owned by System Account. \
+                                     Add `--allow-non-system-account-recipient` to complete the transfer. \
+                                    ".into());
+            }
+            else{
+                false
+            }
+        }
```

Comment:
This whole bit is going to be tough to parse in the future, so how about we add a few more variables in the middle to make this clearer?  Here's what I'm thinking:

```
let recipient_account_info = config
            .rpc_client
            .get_account_with_commitment(&recipient,config.rpc_client.commitment())?
            .value
            .map(|account| {
                (account.owner == config.program_id && account.data.len() == Account::Len, account.owner == system_program::id())
            });
        if let Some((recipient_is_token_account, recipient_is_system_account)) {
            if !recipient_is_token_account && !recipient_is_system_account && !allow_non_system_account_recipient { return Err(...); } 
        } else if !allow_unfunded_recipient {
            return Err(...);
        }
       recipient_account_info.map(|(recipient_is_token_account, _)| recipient_is_token_account).unwrap_or(false)
```